lib/ack_tracker.h

```diff
@@ -84,4 +84,19 @@ ack_tracker_manage_msg_ack(AckTracker *self, LogMessage *msg, AckType ack_type)
   self->manage_msg_ack(self, msg, ack_type);
 }
 
+static inline void
+_ack_tracker_update_source_when_ack_received(LogSource *source, AckType ack_type)
+{
+  if (ack_type == AT_SUSPENDED)
+    {
+      if (log_source_free_to_send(source))
+        log_source_flow_control_suspend(source);
+    }
+  else
+    {
+      if (!log_source_free_to_send(source))
+        log_source_wakeup(source);
```

Comment:
The free_to_send() call is a problem here too.
Instead of calling `wakeup` here, I would reset the suspend bit in `_flow_control_window_size_adjust()`, where we can do it safely.